### PR TITLE
Support definition for resource-method-name

### LINKED ISSUE #42220: [Bug]: The resource paths are highlighted for "Go to definition" instead of the method
### Description

In a resource access, "Go to definition" should be highlighted on the method rather than the resource access parameters. However, as shown in the following example, it is highlighted on the resource parameter.

image

### Steps to Reproduce

Attempt to use "Go to definition" on the following cursor positions.

```ballerina
import ballerina/http;

public function main() returns error? {
    http:Client cl = check new ("http://localhost:9090");
    _ = check cl->/fo<CURSOR1>od/apple.g<CURSOR2>et(targetType = json);
}
```

### Affected Version(s)

Ballerina 2201.8.5 (Swan Lake Update 8)

### OS, DB, other environment details and versions

OS: macOS 14.2.1 23C71
JDK: openjdk 17.0.8 2023-07-18

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose

Fixes #42220

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SymbolFinder.java [hunk 1]
```diff
@@ -1684,13 +1684,16 @@ public void visit(BLangInvocation.BLangResourceAccessInvocation resourceAccessIn
...
             return;
         }
 
         if (this.symbolAtCursor == null) {
-            setEnclosingNode(resourceAccessInvocation.symbol, resourceAccessInvocation.resourceAccessPathSegments.pos);
+            setEnclosingNode(symbol, resourceAccessInvocation.resourceAccessPathSegments.pos);
+        }
+        if (this.symbolAtCursor == null) {
+            setEnclosingNode(symbol, resourceAccessInvocation.name.pos);
         }
```
Comment (nipunayf) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SymbolFinder.java:1697]: Both conditions are the same; hence, the first logic has become redundant.

**Review by nipunayf — COMMENTED** [review]
Also, some of the semantic api tests are failing as well. May need to check those

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SymbolFinder.java (hunk 1 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SymbolFinder.java]: Fixed

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SyntaxNodeToLocationMapper.java [hunk 2]
```diff
@@ -187,7 +187,10 @@ public Optional<Location> transform(RestParameterNode restParameterNode) {
 
     @Override
     public Optional<Location> transform(ClientResourceAccessActionNode clientResourceAccessActionNode) {
-        return Optional.of(clientResourceAccessActionNode.slashToken().location());
+        if (clientResourceAccessActionNode.methodName().isEmpty()) {
+            return Optional.of(clientResourceAccessActionNode.slashToken().location());
+        }
+        return Optional.of(clientResourceAccessActionNode.methodName().get().location());
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SyntaxNodeToLocationMapper.java:193]:
```
public Optional<Location> transform(ClientResourceAccessActionNode clientResourceAccessActionNode) {
    return clientResourceAccessActionNode.methodName()
        .map(MethodName::location)
        .orElseGet(() -> clientResourceAccessActionNode.slashToken().location());
}
```

Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SyntaxNodeToLocationMapper.java:193]:
```suggestion
    return clientResourceAccessActionNode.methodName()
        .map(MethodName::location)
        .orElseGet(() -> clientResourceAccessActionNode.slashToken().location());

```

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SyntaxNodeToLocationMapper.java [hunk 3]
```diff
@@ -187,7 +187,10 @@ public Optional<Location> transform(RestParameterNode restParameterNode) {
 
     @Override
     public Optional<Location> transform(ClientResourceAccessActionNode clientResourceAccessActionNode) {
-        return Optional.of(clientResourceAccessActionNode.slashToken().location());
+    return clientResourceAccessActionNode.methodName()
+        .map(MethodName::location)
+        .orElseGet(() -> clientResourceAccessActionNode.slashToken().location());
+
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SyntaxNodeToLocationMapper.java:193]:
```suggestion
        if (clientResourceAccessActionNode.methodName().isEmpty()) {
            return Optional.of(clientResourceAccessActionNode.slashToken().location());
        }
        return Optional.of(clientResourceAccessActionNode.methodName().get().location());

```

Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SyntaxNodeToLocationMapper.java:193]:
```suggestion
        if (clientResourceAccessActionNode.methodName().isEmpty()) {
            return Optional.of(clientResourceAccessActionNode.slashToken().location());
        }
        return Optional.of(clientResourceAccessActionNode.methodName().get().location());

```

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SyntaxNodeToLocationMapper.java [hunk 4]
```diff
@@ -187,7 +187,11 @@ public Optional<Location> transform(RestParameterNode restParameterNode) {
 
     @Override
     public Optional<Location> transform(ClientResourceAccessActionNode clientResourceAccessActionNode) {
-        return Optional.of(clientResourceAccessActionNode.slashToken().location());
+        if (clientResourceAccessActionNode.methodName().isEmpty()) {
+            return Optional.of(clientResourceAccessActionNode.slashToken().location());
+        }
+        return Optional.of(clientResourceAccessActionNode.methodName().get().location());
+
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SyntaxNodeToLocationMapper.java:194]:
```suggestion
```

**Review by gimantha — APPROVED** [review]

**Pull request merged**